comms/src/peer_manager/peer_storage.rs

```diff
@@ -345,6 +345,42 @@ where DS: KeyValStore
         }
     }
 
+    /// Check if a specific node_id is in the network region of the N nearest neighbours of the region specified by
+    /// region_node_id
+    pub fn in_network_region(
+        &self,
+        node_id: &NodeId,
+        region_node_id: &NodeId,
+        n: usize,
+    ) -> Result<bool, PeerManagerError>
+    {
+        let mut dists: Vec<NodeDistance> = Vec::new();
+        self.peers
+            .for_each::<PeerKey, Peer, _>(|pair| {
+                let (_, peer) = pair.unwrap();
+                if !peer.is_banned() {
+                    dists.push(region_node_id.distance(&peer.node_id));
+                }
+            })
+            .map_err(PeerManagerError::DatabaseError)?;
+        // Use all available peers up to a maximum of N
+        let max_available = min(dists.len(), n);
```

Comment:
Suggestion. Right now, this is O(n + Nn)

I think we can make it O(nN) in the worst case and eps.O(nN) on average:

First allocate a Vec of max_available distances with large values.
Track the closest N distances found so far in order
When looping through peers above: 
    Calc distance
    If it's in the Top N (i.e. distance < worst closest), insert the new candidate in it's correct slot and drop off the worst.
  If the new Nth closest distance is less than `region_node_id.distance(node_id)`, you can immediately return false without searching further.

Reply:
Change applied